[DirectX][DXIL] Design document for TableGen Spec of DXIL Operations

bharadwajy opened the pull request:
Add an initial design document for TableGen specification of DXIL Operations.

Issue #84932

Commit:
[DirectX][DXIL] Design document for TableGen Spec of DXIL Operations

Add a design document that outlines the choices considered and current
implementation.

hekota commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,207 @@
+primarily, to facilitate lowering of new LLVM intrinsics for HLSL functions
+being added in the front end. It serves to expose any additional considerations
+necessary for an improved design of ``DXIL.td``. The current plan is to design
+and implement **Opton 4** to improve readability and maintainablity while
```
typo - "Option 4"

bharadwajy replied:
Fixed. Thanks!

Commit:
Fix typo

llvm-beanz reviewed (commented):
My $0.02 on reading the content here is that options 2 and 3b don't really accomplish our key goal of making the tablegen a source of truth for DXIL specification.

With option 2 we need to have an external tool that validates the tablegen, which means the tablegen files themselves can't be treated as the source of truth.

For option 3b, again, the tablegen isn't really a clear source of truth. I think having an exclusion list is fragile (what happens if we add a new type), and much less intuitive because nothing about the op declaration clearly conveys that these are the disallowed types rather than the allowed types you just have to know it.

Option 4 seems confusing to me. LLVM's Attr class is for defining IR attributes. I'm not sure how or why we would use attributes for defining the valid overloads.

IIUC, the only real difference between option 1 and option 3a, is that option 3a uses types that can map to one or more type and type matching rather than an array of explicit overloads.

If option 3a is expressive enough to cover all our use cases, that seems fine to me. It also would be trivial to allow a hybrid solution where `OpTypes` can either be an array of type specifiers or an array of arrays of type specifiers. That would actually allow solutions 1 and 3a to be used interchangeably or even in mixed capacity. Like:

```
def SillyNewDXOp  : DXILOpMapping<13, unary, int_dx_some_new_intrinsic,
                         "Returns magic and rainbows!",
                         [[llvm_halforfloat_ty, LLVMMatchType<0>], [llvm_float_ty, llvm_double_ty]]>;
```

That would allow us the flexibility to simplify where reasonable but still have explicit control where we need it.

I have a strong preference for either option 1, 3a, or a hybrid approach. I think that is the most straight forward and simple. Curious if @bogner has other thoughts.

bharadwajy commented:
> If option 3a is expressive enough to cover all our use cases, that seems fine to me. It also would be trivial to allow a hybrid solution where `OpTypes` can either be an array of type specifiers or an array of arrays of type specifiers. That would actually allow solutions 1 and 3a to be used interchangeably or even in mixed capacity. Like:
> 
> ```
> def SillyNewDXOp  : DXILOpMapping<13, unary, int_dx_some_new_intrinsic,
>                          "Returns magic and rainbows!",
>                          [[llvm_halforfloat_ty, LLVMMatchType<0>], [llvm_float_ty, llvm_double_ty]]>;
> ```
>

Is the return type list as shown in the above sample intended to demonstrate the list of all valid overload return types of `SillyNewDXOp`? If so, per my understanding, (a) such a list would be a list of fixed types and (b) `[llvm_halforfloat_ty, LLVMMatchType<0>]` would expand to `[llvm_halforfloat_ty, llvm_halforfloat_ty]`, which doesn't seem to convey additional information. I am assuming that no DXIL operations will return more than one value.

The intent behind defining the more precise overload types `llvm_halforfloat_ty` and `llvm_i16ori32_ty` is to abstract the overload type specification of `[llvm_half_ty, llvm_float_ty]` and `[llvm_i16_ty, llvm_i32_ty]` respectively. A parameter type that is expected to match a previously specified overload type (including these) can be specified via `LLVMMatchType<>`. 

It appears that very few additional precise overload types need to be defined. For example, a common overload type that abstracts an overload of `[llvm_half_ty, llvm_float_ty, llvm_i16_ty, llvm_i32_ty]`, can be defined as `llvm_halforfloatori16ori32_ty`.

> That would allow us the flexibility to simplify where reasonable but still have explicit control where we need it.

I think the specification methodology of option 3a provides the required flexibility and readability.

bogner commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,207 @@
+Another variant of the Option 3a is to specify an exclusion list. An
+exclusion list instead of an override list provides a list of fixed types
+not valid for an DXIL Op and thus need to be excluded from a valid overload
+type lsit of LLVM Intrinsic. The benefits and downsides of this are the same
```
typo: `s/lsit/list/`

bogner commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,207 @@
+``Attr`` records, distinguished via ``AttrProperty`` to model precise types
+as needed. This can provide the necessary readability and the richness of
+specification. Additionally, the other properties of a DXIL Op (such as the
+``bool is_*``) can also be uniformly represented as ``Attr`` records.
```
Why use Attributes.td properties for this rather than a custom field that's specific to DXIL.td? Is there any actual overlap with the existing attrs?

Commit:
Update following feedback.

Commit:
Add details of option that levearges existing DXIL Op class properties

llvm-beanz commented:
Can you include in the examples for each of the options how we would represent the requirements for specific overloads? For example the `Sin` DXIL Op is valid for `float` in SM 6.0+, and valid for `half` in SM 6.2+.

Commit:
Add a section describing Shader Model based constraint specification.

bharadwajy commented:
> Can you include in the examples for each of the options how we would represent the requirements for specific overloads? For example the `Sin` DXIL Op is valid for `float` in SM 6.0+, and valid for `half` in SM 6.2+.

Added a section with the details. Thanks!

llvm-beanz reviewed (commented):
Based on the latest comments and updates. It seems like we need some variation of option 1 or 3 with a way to represent an overload as not just parameters but also some validation.

@bharadwajy, can you do a cleanup pass on the document to reflect either the single planned direction or the narrower list of options we're still considering?

Commit:
Update the details of the currently implemented design.
Delete text describing options not pursued.

bharadwajy commented:
> Based on the latest comments and updates. It seems like we need some variation of option 1 or 3 with a way to represent an overload as not just parameters but also some validation.
> 
> @bharadwajy, can you do a cleanup pass on the document to reflect either the single planned direction or the narrower list of options we're still considering?

Updated the document. Implementation of the design described in the current version of this document is in PR #87803.

llvm-beanz commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,224 @@
+    def Sin  : DXILOpMapping<13, int_sin,
+                             [DXILOpOverload<SM6_3, [llvm_half_ty, llvm_float_ty]>,
+                              DXILOpOverload<SM6_0, [llvm_float_ty]>],
+                             "Returns sine(theta) for theta in radians.">;
```
There is one more odd availability case that this notation doesn't handle (although I suspect it can be extended to do so). The `ddx` intrinsic is a good example:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-ddx

`ddx` was added for pixel shaders in SM 2.x, but was not added to compute shaders until SM 6.6:

https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Derivatives.html

The overload is the same, but the availability is by-stage and changes in later versions.

bharadwajy replied:
> There is one more odd availability case that this notation doesn't handle (although I suspect it can be extended to do so). The `ddx` intrinsic is a good example:
...
> `ddx` was added for pixel shaders in SM 2.x, but was not added to compute shaders until SM 6.6:
...
> The overload is the same, but the availability is by-stage and changes in later versions.

Thanks! I plan to extend the specification ability to include Shader Kind as another refinement criterion in a (near-term) future PR.

bogner commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,224 @@
+#. ``ops`` - list of operands encapsulating the index and valid (fixed or overload) types
+#. ``oload_types`` - Valid overload types of the DXIL op
+#. Rest of the attributes represented using ``is_*`` booleans along with
+   ``shader_model_translated`` and ``shader_stages``
```
I find this list a bit confusing. You're trying to explain why some of the fields from the `hctdb.py` field descriptions above are necessary, but it comes off kind of like you're just repeating what the fields are for.

It'd probably be better to phrase this in terms of what we need in DXIL.td and refer back to how that covers the parts of `hctdb.py`'s implementation instead.

> To fully represent the needs of the DXIL Op mapping in LLVM we'll need the following:
> 
> 1. The DXIL operation ID (``dxil_opid`` from ``hctdb.py``)
> 2. The generic or HLSL-specific intrinsic that we're mapping from
> 3. The "operation class" for the op - that is, the name and function signature such that we can construct a function of the ``dx.op.<class-name>.<overload>`` kind that DXIL expects.
> 4. Constraints on overloads, so that we can avoid generating invalid operations that happen to match the operation class
> 5. Constraints on availability per shader model
> 6. Constraints on availability per shader stage (Note: not yet handled by this spec)
> 7. Constaints on XYZ

It would be nice to fill out the list of constraints we'll need to handle for the other `is_*` booleans in `hctdb.py` for completeness, though I am comfortable getting this in with notes saying where this will continue to evolve for now.

Commit:
Update Major and Minor version representation in ShaderModel class

llvm-beanz reviewed (commented):
My only suggested change here is to create DXIL Sub-Architectures to represent DXIL versions (SPIR-V does this already), then we can key DXIL.td off the DXIL version instead of the shader model version.

@bogner & @farzonl, do either of you have any feedback on this?

Commit:
Delete discarded options and flesh out the proposed design option.
Also incorporate PR suggestion to add descriptions of all properties
of DXIL Operations.

hekota commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,207 @@
+  * "Standard" LLVM intrinsics (e.g., ``llvm.sin.*``) and
+  * HLSL intrinsics (defined as LLVM intrinsics in ``llvm/include/llvm/IR/IntrinsicsDirectX.td``, e.g., ``llvm.dx.*``)
+
+  These are  collectively referred to as `LLVM Intrinsics` in this note.
```
The rendering still seems off. Have you considered merging the two lists together?

#. Using LLVM instructions (standard LLVM intrinsics, e.g., ``llvm.sin.*``)
#. Using LLVM External functions, also called HLSL intrinsics. These are defined in ``llvm/include/llvm/IR/IntrinsicsDirectX.td`` (e.g., ``llvm.dx.*``).

Commit:
Fix rendering per PR feedback

llvm-beanz commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,179 @@
+      class OpAttributes;
+
+   - memory access - ``ReadNone``, ``ReadNone``
+   - ``IsDerivative``, ``IsGradient``, ``IsFeedback``, ``IsWave``, ``NeedsUniformInputs``, ``IsBarrier``
```
As we discussed this morning I'm unsure these properties are required.

llvm-beanz commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,179 @@
+     def Cos  : DXILOpProperties<12, int_cos, [llvm_half_ty, llvm_float_ty],
+                                "Returns cosine(theta) for theta in radians.">;
+     def Sin  : DXILOpProperties<13, int_sin, [llvm_half_ty, llvm_float_ty],
+                           "Returns sine(theta) for theta in radians.">;
```
I'm curious what the representation looks like for a more exotic DXIL Op. Like maybe `ddx`? How do we represent that `ddx` is available in SM 2.0+ for Pixel shaders and SM 6.6+ for Compute, Mesh and Amplification?

How do we handle `RawBufferLoad` where the `double` and `i64` overloads are only valid for SM 6.3+?

bharadwajy replied:
> I'm curious what the representation looks like for a more exotic DXIL Op. Like maybe `ddx`? How do we represent that `ddx` is available in SM 2.0+ for Pixel shaders and SM 6.6+ for Compute, Mesh and Amplification?
> 
> How do we handle `RawBufferLoad` where the `double` and `i64` overloads are only valid for SM 6.3+?

Thanks for flagging these. I've noted this and plan to update the design spec in subsequent PRs, if needed, as I make progress with implementation and learn about the specific requirements to represent DXIL ops that may not be sufficiently represented using the current specification mechanism.

Commit:
Focus design on specification of DXIL Op properties required for
DXIL backend only.

bogner commented on llvm/docs/DirectX/DXILOpTableGenDesign.rst:
```diff
@@ -0,0 +1,160 @@
+
+   1. Required minimum Shader Model (``shader_model``).
+   2. Minimum shader model required with translation by linker (``shader_model_translated``)
+   3.  List of shader stages applicable to (``shader_stages``), empty for all.
```
I feel like the valid shader stages is as relevant to the backend as the required DXIL version (ie, for validation that we're generating correct DXIL).

bogner reviewed (approved):
As Chris mentioned in https://github.com/llvm/llvm-project/pull/85170#discussion_r1618007508, there are a few glaring gaps here that I think will need to be resolved in order to generate correct DXIL:

1. Derivatives, which became valid in more shader stages in SM6.6
2. Atomics, which gained 64 bit overloads in SM6.6
3. Sample ops gained [integer overloads](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_7_Advanced_Texture_Ops.html#integer-sampling) in SM6.7
4. Sample and Texture ops may be affected by SM6.7 [programmable offsets](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_7_Advanced_Texture_Ops.html#programmable-offsets)

It isn't ideal to move forward with this without at least having an idea of how we'll handle those scenarios. However, it's probably sensible to get this in tree and iterate on it there as the design evolves.

jayfoad commented:
@bharadwajy this seems to break the docs build with:
```
[1/2] Generating man Sphinx documentation for llvm into "/home/runner/work/llvm-project/llvm-project/llvm-build/docs/man"
FAILED: docs/CMakeFiles/docs-llvm-man /home/runner/work/llvm-project/llvm-project/llvm-build/docs/CMakeFiles/docs-llvm-man 
cd /home/runner/work/llvm-project/llvm-project/llvm-build/docs && /usr/local/bin/cmake -E env /opt/hostedtoolcache/Python/3.11.9/x64/bin/sphinx-build -b man -d /home/runner/work/llvm-project/llvm-project/llvm-build/docs/_doctrees-llvm-man -q -t builder-man -D version=19 -D release=19.0.0git -tPreRelease -W /home/runner/work/llvm-project/llvm-project/llvm/docs /home/runner/work/llvm-project/llvm-project/llvm-build/docs/man

Warning, treated as error:
/home/runner/work/llvm-project/llvm-project/llvm/docs/DirectX/DXILOpTableGenDesign.rst:document isn't included in any toctree
```
See for example: https://github.com/llvm/llvm-project/actions/runs/9306373243/job/25615439087?pr=93801

bharadwajy commented:
> @bharadwajy this seems to break the docs build with:
> 
> ```
> [1/2] Generating man Sphinx documentation for llvm into "/home/runner/work/llvm-project/llvm-project/llvm-build/docs/man"
> FAILED: docs/CMakeFiles/docs-llvm-man /home/runner/work/llvm-project/llvm-project/llvm-build/docs/CMakeFiles/docs-llvm-man 
> cd /home/runner/work/llvm-project/llvm-project/llvm-build/docs && /usr/local/bin/cmake -E env /opt/hostedtoolcache/Python/3.11.9/x64/bin/sphinx-build -b man -d /home/runner/work/llvm-project/llvm-project/llvm-build/docs/_doctrees-llvm-man -q -t builder-man -D version=19 -D release=19.0.0git -tPreRelease -W /home/runner/work/llvm-project/llvm-project/llvm/docs /home/runner/work/llvm-project/llvm-project/llvm-build/docs/man
> 
> Warning, treated as error:
> /home/runner/work/llvm-project/llvm-project/llvm/docs/DirectX/DXILOpTableGenDesign.rst:document isn't included in any toctree
> ```
> 
> See for example: https://github.com/llvm/llvm-project/actions/runs/9306373243/job/25615439087?pr=93801

@jayfoad Thanks for letting me know of this. I have a [change](https://github.com/llvm/llvm-project/compare/main...bharadwajy:llvm-project:dxil_td/design-doc-toctree) that I believe would fix this failure. However, I am unable to verify it on my dev machines as I see the following failure.

```
Warning, treated as error:
cannot cache unpickable configuration value: 'myst_heading_slug_func' (because it contains a function, class, or module object)
```

I have not found a way to address this issue after a quick internet search. Would you have any suggestions?

bharadwajy commented:
> I have not found a way to address this issue after a quick internet search. Would you have any suggestions?

Never mind. I temporarily added `"config.cache"` to `suppress_warnings` in `conf.py` to verify the fix.
